mirgecom/limiter.py

```diff
@@ -0,0 +1,108 @@
+"""
+:mod:`mirgecom.limiter` is for limiters and limiter-related constructs.
+
+Field limiter functions
+^^^^^^^^^^^^^^^^^^^^^^^
+
+.. autofunction:: bound_preserving_limiter
+.. autofunction:: cell_characteristic_size
+
+"""
+
+__copyright__ = """
+Copyright (C) 2022 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from grudge.discretization import DiscretizationCollection
+import grudge.op as op
+
+
+def cell_characteristic_size(actx, dcoll: DiscretizationCollection):
+    r"""Evaluate cell area or volume."""
+    zeros = actx.thaw(actx.freeze(dcoll.nodes()))[0]
+    return op.elementwise_integral(dcoll, zeros + 1.0)
+
+
+def bound_preserving_limiter(dcoll: DiscretizationCollection, cell_size, field,
+                                  mmin=0.0, mmax=None, modify_average=True):
+    r"""Implement a slope limiter for bound-preserving properties.
+
+    The implementation is summarized in [Zhang_2011]_, Sec. 2.3, Eq. 2.9,
+    which uses a linear scaling factor of the high-order polynomials.
+
+    By default, the average is also bounded to ensure that the solution is
+    always positive. This option can be modified by an boolean argument.
+    In case the average becomes negative, the limiter will drop to order zero
+    and may not fix the negative values. This operation is not described in
+    the original reference but greatly increased the robustness.
+
+    An additional argument can be added to bound the upper limit.
+
+    Parameters
+    ----------
+    dcoll: :class:`grudge.discretization.DiscretizationCollection`
+        Grudge discretization with boundaries object
+    cell_size: meshmode.dof_array.DOFArray or numpy.ndarray
+        A field or collection of scalar fields to limit
```

Comment:
I'm having a little trouble understanding this parameter's doc. Could the doc for this just be something like: "The cell volumes" ? 

Is there a particular reason that the cell volumes should be passed in instead of computed down inside this function?

Reply:
My bad, just copied-and-paste and didn't change. Thanks for pointing that out.

I was computing the cell size/volume at the driver, since it doesn't change in time, and passing to the limiter function.. However, if that shouldn't matter, we can compute down inside the function.